### PR TITLE
osdc/ObjectCacher: overwrite might cause stray read request callbacks

### PR DESCRIPTION
**Pull request by dillaman**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by trociny — APPROVED** [review]
LGTM

File: src/librbd/image/OpenRequest.cc [hunk 1]
```diff
@@ -559,7 +559,8 @@ Context* OpenRequest<I>::handle_init_plugin_registry(int *result) {
 template <typename I>
 Context *OpenRequest<I>::send_init_cache(int *result) {
   // cache is disabled or parent image context
-  if (!m_image_ctx->cache || m_image_ctx->child != nullptr) {
+  if (!m_image_ctx->cache || m_image_ctx->child != nullptr ||
+      !m_image_ctx->data_ctx.is_valid()) {
```
[the remark this answers is not in the document]
Comment (dillaman) [reply src/librbd/image/OpenRequest.cc:562]: Done

**Comment by batrick** [comment]
Passes cephfs testing, good to merge from our side.

**Comment by batrick** [comment]
https://pulpito.ceph.com/?branch=wip-pdonnell-testing-20200926.000836

failures known/unrelated.

**Pull request merged**